### ISSUE TITLE
Multiple base classes to choose from for Copies

### ISSUE THREAD

**Issue by CosmoMyzrailGorynych** [ct-js/ct-js-old#448]
## TL;DR

Multiple base classes for templates means that a user is able to select whether they want to base their template on top of PIXI.AnimatedSprite (pre-ct.js v4 behavior) or another PIXI.DisplayObject subclass.

**Use cases include:**

* Using existing pixi.js classes: 9-slice planes, text labels and such as bases for ct.js templates, so a user has a more WYSIWYG way working with such classes.
* Creating more complex classes on top of pixi.js, mainly to create UI elements: buttons, input fields, key selectors, joysticks and such.
* 3D 👀

This is also a nice preparation to pixi.js v8 which removes `this.children` functionality from everything but `PIXI.Container`.

The base classes feature consists of:

* ct.IDE changes in templates-editor: users should be able to choose from several base classes and be able to switch them afterwards.
* ct.js client library changes to support various base classes:
  * Universal `Copy` mixin to add ct.js functionality to various base classes;
  * modifications to the `rooms` and `templates` API to create new copies of various base classes.
* Room editor tools to support WYSIWYG design workflow.
* Documentation and tutorial updates to reflect these changes: FattyCat, Space Shooter, and Platformer tutorials can use dynamic text labels instead of using `PIXI.Text` class directly in ct.js code.

![Image](https://github.com/ct-js/ct-js/assets/3174357/664b5a00-fd03-40bb-be35-39c6de69421e)

![Image](https://github.com/ct-js/ct-js/assets/3174357/89ca45b3-985b-4bb3-8199-12b240c58d00)

## TODO:

- [x] Make a Copy mixin
- [x] Make TypeScript believe that all this is valid
- [x] Room editor should support three starting base classes:
  - [x] AnimatedSprite
  - [x] Text
  - [x] NineSlicePlane 
- [x] The client lib should implement base classes and UI configs from ct.IDE for them:
  - [x] AnimatedSprite
  - [x] Text
  - [x] NineSlicePlane 
- [x] Code editors should respect the picked base class, exposing additional properties when appropriate
- ~~Add skeletal animation as a fourth base class instead of adding it as a child for AnimatedSprite~~
  - ~~Need to complete `res.makeSkeleton` first.~~
  - Can't. Including a Spine runtime would make ct.js not fully open-source.
- [x] Event list should filter out `PIXI.AnimatedSprite` events when a different base class is specified. Ideally there should be a way to specify required base class in the modded/built-in event.
  - [x] And now behaviors' event list should warn users that this or that event cannot be applied to everything.
  - ~~And the exporter should filter out invalid events as well.~~ No easy way to do it with dynamic behaviors.
- [x] UI: anchor buttons work but don't get their `active` class applied. [Video](https://cdn.discordapp.com/attachments/490054452926218240/1170557913924911194/2023-11-04_22-58-31.mp4?ex=65597a02&is=65470502&hm=5e0c4cb1b85f43e55fe21995017b1911539a9d142d86ec4d1f0ab7d6a82fd2a1&)
- [x] Collision shapes:
  - [x] For text labels
  - ~~For skellies~~
  - [x] For 9-slice panels
- [ ] Doc updates:
  - [x] A page about various base classes, or maybe augment the page of the Copy class.
  - [x] Remove AnimatedSprite-specific fields from the Copy page.
  - [ ] Update tutorials.
- [x] Update examples, mainly redo text labels.

**Comment by CosmoMyzrailGorynych** [comment]
Closed in 805190b9029ddde5bd3b4c043c666f14ba0be06f
